### ISSUE TITLE
TypeError: Unable to cast function/method to java type 'java.lang.Runnable'

### ISSUE THREAD

**Issue by astrelsky** [jpype-project/jpype#1051]
```python
import jpype
import jpype.imports

class MyClass:

    def fun(self):
        print("this doesn't")

def fun():
    print('this works')

if __name__ == '__main__':
    jpype.startJVM()
    from java.lang import Runnable
    mc = MyClass
    cb = Runnable @ fun
    cb.run()
    cb = Runnable @ mc.fun
    cb.run()
```

I'm going to assume this is caused by #1017. This is not a problem in jpype 1.3.0.

**Comment by Thrameos** [comment]
Yes.  And that is why I didn't think trying to check arguments was a good idea given there is no API in Python to do so.  The issue here is that the ``__code__`` object does not know if the method is bound or not.   Thus checking the number of arguments in

However, your example is definitely buggy.   MyClass.fun required one argument so it should not be castable to  Runnable.  MyClass().fun on the other hand should work but doesn't..  I will have to replace the implementation from #1017.

**Comment by astrelsky** [comment]
> Yes.  And that is why I didn't think trying to check arguments was a good idea given there is no API in Python to do so.  The issue here is that the ``__code__`` object does not know if the method is bound or not.   Thus checking the number of arguments in order to reject the cast causes errors.   
> 
> However, your example is definitely buggy.   MyClass.fun required one argument so it should not be castable to  Runnable.  MyClass().fun on the other hand should work but doesn't..  I will have to replace the implementation from #1017. 

The result is a callable python object equivalent of `lambda a: a.fun()`

If I remember correctly `inspect` allows runtime static inspection of python objects.

**Comment by astrelsky** [comment]
> Yes. And that is why I didn't think trying to check arguments was a good idea given there is no API in Python to do so. The issue here is that the `__code__` object does not know if the method is bound or not. Thus checking the number of arguments in order to reject the cast causes errors.
> 
> However, your example is definitely buggy. MyClass.fun required one argument so it should not be castable to Runnable. MyClass().fun on the other hand should work but doesn't.. I will have to replace the implementation from #1017.

You can use [PyMethod_Check](https://docs.python.org/3/c-api/method.html#c.PyMethod_Check) to determine if it is a bound or not. I will hold off on checking #1054 to ensure it is fixed until tomorrow morning.

**Comment by Thrameos** [comment]
Yes.  I used PyMethod_Check.   There is only one set of cases that we haven't addressed here.   And that is if someone were to pass a JPype method or function.   I suppose those should be tested as well.
